internal/terraform/transform_destroy_edge.go

```diff
@@ -337,16 +337,6 @@ func (t *pruneUnusedNodesTransformer) Transform(g *Graph) error {
 					for _, v := range g.UpEdges(n) {
 						switch v.(type) {
 						case graphNodeExpandsInstances:
-							// Root module output values (which the following
-							// condition matches) are exempt because we know
-							// there is only ever exactly one instance of the
-							// root module, and so it's not actually important
-							// to expand it and so this lets us do a bit more
-							// pruning than we'd be able to do otherwise.
-							if tmp, ok := v.(graphNodeTemporaryValue); ok && !tmp.temporaryValue() {
```

Comment:
@jbardin, how important is this optimisastion within the overall context of performance issues we've been facing? 

Essentially since we can now refer to the expander during the apply phase in the evaluator (as outputs are being processed properly and can refer to deferred resources) we do need to expand resource nodes even if the only thing referring to them is a root output. This is simply because the expander must be aware of resources or it panics.

This optimisation meant that resource expansion nodes were being removed even though they were being referenced and we start getting panics. Since we want outputs to behave a specific way when referencing resources that are being deferred we do need to check for partial expansions during the apply.

Reply:
@nfagerlund, hopefully this question for James adds some context to what is happening here!

Reply:
The `pruneUnusedNodesTransformer` isn't an optimization, but rather helps handle the case of complex destroy scenarios (see #25018 where this transformer last had some major refactoring). The conversation around the block being removed may be interesting too [starting here](https://github.com/hashicorp/terraform/pull/31395#issuecomment-1190886303).

The gist of the process is that "temporary" values, i.e. ones not written to the state, are always evaluated again during apply, but because apply can also be destroying portions of the state (or portions are already missing entirely), the config may be inserting some nodes which cannot be evaluated because their dependencies have already been destroyed, and will otherwise error out if they are not pruned.

Reply:
Maybe the failures that Martin had in the second link there shed some light on why this was needed. My hunch is that the case we still need to verify here is: a local value referenced by a root output, which references a resource that is already destroyed.

Reply:
I wondered if a potential behaviour change here was to simply stop outputs from evaluating during destroy only plans? I think that was another potential workaround I had for this issue.

Reply:
The same hold during apply too, but another problem is that you might not have a destroy-only plan, it could be a normal plan where portions of the configuration have been removed.

Reply:
Would that not lead to a validation error? An output referring to a part of the configuration that has been removed?

Reply:
that's what I get for commenting on the way out the door last night, and was only thinking about the resources. Yeah, any output would need to come from valid configuration, even during destroy. The reason they were always there in general is that originally all `apply` steps were supposed to be equal, with no knowledge of which command was being run. We've since accepted that applying a destroy plan is different and different actions need to be taken, so perhaps we can just remove the root outputs in that case.

Reply:
@nfagerlund, based on this I think we need to reintroduce the check here that I deleted, and reintroduce the relevant test. This will then likely lead to some panics happening in some tests. 

To fix that, we can come into https://github.com/hashicorp/terraform/blob/main/internal/terraform/node_output.go#L367 and check that if we're an output in the root module and `n.DestroyApply` is true then just return `cty.NullValue(cty.DynamicPseudoType)` for this output on the basis that this is a destroy action so the outputs are going to be null once everything is destroy anyway.

Feel free to ping me or anything if I'm not making any sense!

Reply:
This took some digging to fully understand, but the eventual outcome is: 

- The core logic of this PR added new spots where we expect resource expand nodes to have properly executed before trying to reference values. 
- The prune step was violating that expectation by pruning expand nodes in apply graphs. 
- But pruning ONLY serves a useful purpose in destroy runs; there's genuinely no value to running it in a non-destroy context. We were only doing that in order to try and treat destroy and non-destroy the same, which we've now recognized isn't always viable. 
- So I added a commit to skip the entire prune transform in non-destroy applies.